File: 0834-sum-of-distances-in-tree/0834-sum-of-distances-in-tree.cpp
class Solution {
public:
    vector<int> count, ans;
    void dfs(vector<unordered_set<int>> &adj, int root, int parent) {
        for(int node : adj[root]) {
            if(node == parent) continue;
            dfs(adj, node, root);
            count[root] += count[node];
            ans[root] += ans[node] + count[node];
        }
    }

    void dfs2(vector<unordered_set<int>> &adj, int root, int parent) {
        for(int node : adj[root]) {
            if(node == parent) continue;
            ans[node] = ans[root] - count[node] + count.size() - count[node];
            dfs2(adj, node, root);
        }
    }

    vector<int> sumOfDistancesInTree(int n, vector<vector<int>>& edges) {
        count.resize(n, 1);
        ans.resize(n, 0);

        vector<unordered_set<int>> adj(n);
        for(auto edge : edges) {
            adj[edge[0]].insert(edge[1]);
            adj[edge[1]].insert(edge[0]);
        }

        dfs(adj, 0, -1);
        dfs2(adj, 0, -1);
        return ans;
    }
};

File: 0834-sum-of-distances-in-tree/0834-sum-of-distances-in-tree.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[]}
 */

function dfs(adj, root, parent, count, ans) {
    for (let node of adj[root]) {
        if (node === parent) continue;
        dfs(adj, node, root, count, ans);
        count[root] += count[node];
        ans[root] += ans[node] + count[node];
    }
}

function dfs2(adj, root, parent, count, ans) {
    for (let node of adj[root]) {
        if (node === parent) continue;
        ans[node] = ans[root] - count[node] + count.length - count[node];
        dfs2(adj, node, root, count, ans);
    }
}

var sumOfDistancesInTree = function(n, edges) {
    const count = new Array(n).fill(1);
    const ans = new Array(n).fill(0);

    const adj = Array.from({length : n}, () => []);
    for(let edge of edges) {
        adj[edge[0]].push(edge[1]);
        adj[edge[1]].push(edge[0]);
    }
   
    dfs(adj, 0, -1, count, ans);
    dfs2(adj, 0, -1, count, ans);
    return ans;
};